In [1]:
import pandas as pd

In [2]:
data = pd.read_csv('artwork_sample.csv')

In [3]:
## There are many different kind of values here int, float and NaN

In [4]:
data.head()

id accession_number          artist artistRole  artistId  \
0  1035           A00001   Blake, Robert     artist        38   
1  1036           A00002   Blake, Robert     artist        38   
2  1037           A00003   Blake, Robert     artist        38   
3  1038           A00004   Blake, Robert     artist        38   
4  1039           A00005  Blake, William     artist        39   

                                               title                dateText  \
0  A Figure Bowing before a Seated Old Man with h...          date not known   
1  Two Drawings of Frightened Figures, Probably f...          date not known   
2  The Preaching of Warning. Verso: An Old Man En...                 ?c.1785   
3     Six Drawings of Figures with Outstretched Arms          date not known   
4  The Circle of the Lustful: Francesca da Rimini...  1826–7, reprinted 1892   

                                              medium  \
0  Watercolour, ink, chalk and graphite on paper....   
1                                  Graphite on paper   
2        Graphite on paper. Verso: graphite on paper   
3                                  Graphite on paper   
4                            Line engraving on paper   

                                          creditLine    year  acquisitionYear  \
0                Presented by Mrs John Richmond 1922     NaN             1922   
1                Presented by Mrs John Richmond 1922     NaN             1922   
2                Presented by Mrs John Richmond 1922  1785.0             1922   
3                Presented by Mrs John Richmond 1922     NaN             1922   
4  Purchased with the assistance of a special gra...  1826.0             1919   

              dimensions  width  height  depth units  inscription  \
0  support: 394 x 419 mm    394     419    NaN    mm          NaN   
1  support: 311 x 213 mm    311     213    NaN    mm          NaN   
2  support: 343 x 467 mm    343     467    NaN    mm          NaN   
3  support: 318 x 394 mm    318     394    NaN    mm          NaN   
4    image: 243 x 335 mm    243     335    NaN    mm          NaN   

   thumbnailCopyright                                       thumbnailUrl  \
0                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
1                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
2                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
3                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
4                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   

                                                 url  
0  http://www.tate.org.uk/art/artworks/blake-a-fi...  
1  http://www.tate.org.uk/art/artworks/blake-two-...  
2  http://www.tate.org.uk/art/artworks/blake-the-...  
3  http://www.tate.org.uk/art/artworks/blake-six-...  
4  http://www.tate.org.uk/art/artworks/blake-the-...

#### height column: The height of the artwork, medium column: tools used to create the artwork

In [6]:
# we can see their data types by running data.dtypes read csv automatically guessed our data types where 
# object data type is string in our case

In [7]:
data.dtypes

id                      int64
accession_number       object
artist                 object
artistRole             object
artistId                int64
title                  object
dateText               object
medium                 object
creditLine             object
year                  float64
acquisitionYear         int64
dimensions             object
width                   int64
height                  int64
depth                 float64
units                  object
inscription           float64
thumbnailCopyright    float64
thumbnailUrl           object
url                    object
dtype: object

In [8]:
# the most common way to change the column type is astype(nameof type)

In [9]:
data.acquisitionYear  

0    1922
1    1922
2    1922
3    1922
4    1919
5    1919
6    1919
7    1919
8    1919
9    1919
Name: acquisitionYear, dtype: int64

In [10]:
# lets change it to float so we can add half years

In [11]:
data.acquisitionYear.astype(float)

0    1922.0
1    1922.0
2    1922.0
3    1922.0
4    1919.0
5    1919.0
6    1919.0
7    1919.0
8    1919.0
9    1919.0
Name: acquisitionYear, dtype: float64

In [12]:
# if we check the orig col its still int

In [13]:
data.acquisitionYear.dtype

dtype('int64')

In [14]:
# as type does not change the column in place but rather a brand new series of new type

In [15]:
data.acquisitionYear = data.acquisitionYear.astype(float)

In [16]:
# check the data type now

In [17]:
data.acquisitionYear.dtype

dtype('float64')

In [18]:
fulldf = pd.read_csv('artwork_data.csv',low_memory=False) # reading full copy of the dataset

In [19]:
fulldf.head()

id accession_number          artist artistRole  artistId  \
0  1035           A00001   Blake, Robert     artist        38   
1  1036           A00002   Blake, Robert     artist        38   
2  1037           A00003   Blake, Robert     artist        38   
3  1038           A00004   Blake, Robert     artist        38   
4  1039           A00005  Blake, William     artist        39   

                                               title                dateText  \
0  A Figure Bowing before a Seated Old Man with h...          date not known   
1  Two Drawings of Frightened Figures, Probably f...          date not known   
2  The Preaching of Warning. Verso: An Old Man En...                 ?c.1785   
3     Six Drawings of Figures with Outstretched Arms          date not known   
4  The Circle of the Lustful: Francesca da Rimini...  1826–7, reprinted 1892   

                                              medium  \
0  Watercolour, ink, chalk and graphite on paper....   
1                                  Graphite on paper   
2        Graphite on paper. Verso: graphite on paper   
3                                  Graphite on paper   
4                            Line engraving on paper   

                                          creditLine  year  acquisitionYear  \
0                Presented by Mrs John Richmond 1922   NaN           1922.0   
1                Presented by Mrs John Richmond 1922   NaN           1922.0   
2                Presented by Mrs John Richmond 1922  1785           1922.0   
3                Presented by Mrs John Richmond 1922   NaN           1922.0   
4  Purchased with the assistance of a special gra...  1826           1919.0   

              dimensions width height  depth units inscription  \
0  support: 394 x 419 mm   394    419    NaN    mm         NaN   
1  support: 311 x 213 mm   311    213    NaN    mm         NaN   
2  support: 343 x 467 mm   343    467    NaN    mm         NaN   
3  support: 318 x 394 mm   318    394    NaN    mm         NaN   
4    image: 243 x 335 mm   243    335    NaN    mm         NaN   

  thumbnailCopyright                                       thumbnailUrl  \
0                NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
1                NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
2                NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
3                NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
4                NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   

                                                 url  
0  http://www.tate.org.uk/art/artworks/blake-a-fi...  
1  http://www.tate.org.uk/art/artworks/blake-two-...  
2  http://www.tate.org.uk/art/artworks/blake-the-...  
3  http://www.tate.org.uk/art/artworks/blake-six-...  
4  http://www.tate.org.uk/art/artworks/blake-the-...

In [20]:
fulldf.dtypes

id                      int64
accession_number       object
artist                 object
artistRole             object
artistId                int64
title                  object
dateText               object
medium                 object
creditLine             object
year                   object
acquisitionYear       float64
dimensions             object
width                  object
height                 object
depth                 float64
units                  object
inscription            object
thumbnailCopyright     object
thumbnailUrl           object
url                    object
dtype: object

In [21]:
# a lot less data types (more object types than before) 
# because when read csv guesses at the datatypes of the columns if there are bad values in some of the rows
# then it has to set the data type to object.
# we can try to see this by setting the column type to float (height)

In [22]:
fulldf.height.astype(float)

ValueError: could not convert string to float: 'mm'

In [23]:
# we can see atleast one of the rows consists of string mm read csv canot figure out how to convert it into float
# so instead we will use to numeric method from pandas

In [24]:
pd.to_numeric(fulldf.height)

ValueError: Unable to parse string "mm" at position 41339

In [25]:
# we are able to see that it still has trouble parsing the mm row and shows us the position of the error
# so we have to add the param called errors = "coerce" and tell it to coerce the errors into float

In [26]:
pd.to_numeric(fulldf.height, errors="coerce")

0         419.0
1         213.0
2         467.0
3         394.0
4         335.0
          ...  
69196     305.0
69197     305.0
69198    2410.0
69199       NaN
69200     660.0
Name: height, Length: 69201, dtype: float64

In [27]:
# the new series is coerced into float type
# and if we check the original row having the trouble

In [28]:
pd.to_numeric(fulldf.height,errors="coerce")[41339]

nan

In [29]:
# it couldnt guess what the number mm was so it converted it to not a number
# lastly lets check the column type in the data frame and its still an object

In [30]:
fulldf.height.dtype

dtype('O')

In [31]:
# we need to remember, in order to make the changes permenant in that dataframe set that column  with the series
# that gets returned from to numeric method

In [32]:
fulldf.height = pd.to_numeric(fulldf.height, errors="coerce")

In [33]:
# Now check the data type

In [34]:
fulldf.height.dtype

dtype('float64')

# Tasks:

##### The hints and the tips are purposely not so straight forward so that extra attention is paid while solving the task. You can look around the tasks before of after the current task to get the implicit hints as well 

<ol>
<li>Data Aggregation</li>
<li>Data Normalization</li>
<li>Data Transformation</li>
<li>Data Filteration</li>
</ol>

### 1. Data Aggregation
<ol>
<li>Read in artwork_sample.csv</li>
<li>Store "year" min,max,sum,mean,std in a list and print  (NaN values are automatically excluded from agg functions)</li>
    <li>Print the output of what sum() is doing on the artist column</li>
<li>Output min() and max() of artist column and print what is min and max doing for string values also which aggregation function/s won't work here (We can use some aggregation function on strings as well e.g min)</li>

<li>Apply sum() on whole dataframe and explain how sum works across whole data frame (infer if you have to)</li>
<li>Apply .agg function to the dataframe with argument min and print what it does</li>
<li>pass multiple arguments to .agg function as array := min,max,mean,std and print in one line what is happening there 1st col should be [1035, 1044, 1039, 3.02765] explain the NaN's for string type columns</li>    

### 2. Data Normalization (Rescaling the values)
<ol>
<li>Read in artwork_sample.csv in dataframe named "data"</li>
<li>Print in one line which columns would it make sense to apply nomalization HINT: where mean works</li>
<li>Apply Standard Normalization (Around 0) of height column by subtracting the mean and dividing by std dev Output:[0:1.010198,1:-2.072777,2:1.728561, ...] Hint: Column Broadcasting one line of code. </li>
<li>Normalize the height column values between 0 and 1 Hint: minmax = ( height - height.min( ) )/height.what( ) - height.min( ) ) OUTPUT:[0:0.811024,1:0.000000,2:1.000000,...]</li>
<li>Now print the min and the max value of the column you got From previous step to verify min and max values are 0 and 1</li>
<li>Create a column standardized_height and append the  normalized values there Hint:data['standardized_height']= answer column from previous step Tips: Dot notation can't create new columns like data.new_col = answer_col</li>
<li>Output dataframe using head() to show that the new column is there</li>    
</ol>

### 3. Data Transformations
<ol>
<li>Read in artwork_sample.csv in dataframe named "data"</li>
<li>We might want to change the whole column in some way. A transform( ) takes a function to apply on the whole series. Apply transform on height column. Convet height from milimeters (current unit) to centimeters. HINTS: 1==> There are 10mm in 1cm. 2==> data.height.transform( lambda x: x+1 ) outputs the transformed x value. First 3 OUTPUTs: [0:41.9,1:21.3,2:46.7,...]</li>
<li>Transforms are more powerful when joined with group by e.g data.groupby('artist').transform('nunique') counts the number of unique values per row and per column grouped by artist for example if you run the above command it tells you that the first 4 rows where artist name was "Blake,Robert" id = 4 means the 4 works we had by robert blake there were three different mediums used i.e. 3 unique values in the medium column for the author Blake Robert. TASK: Lets get the mean height of an artist's artwork Hint: use the syntax from point number 3 and add ['something'] after the artist. This something is the attribute of an artist on which you want to apply transform on and change the transform accordingly OUTPUT:[0:373.25, 1:373.25, 2:373.25, 3:373.25, 4:337.00, 5:337.00] </li>
<li>Explain the output of the above point in the print statement why repeating values? Hint check the original data frame</li>
<li>Print in one line what was the mean height of the Artist "Blake, William" and is it in mm or cm Hint: Track you work to answer in print statement. OUTPUT:[Mean Height, CM or MM]</li>
<li>Store the mean heights of the artwork in the new column "mean_height_by_artist" in the "data" dataframe and output first 5 records. OUTPUT: A new column "mean_height_by_artist" and [373.25, 373.25, 373.25, 373.25, 337.00] as first 5 rows</li>    
</ol>

### 4. Data Filtering
<ol>
<li>Read Crefully: Read in the entire data set artwork_data.csv in dataframe named "data" avoid the warning using low_memory=False while reading in the csv</li>
<li>There are too many columns let's reduce the output to two columns using .filter function by passing an array of coulumns we want i.e. id and artist and reduce the output to first five rows</li>
    <li>Tasks for Home:</li>
    <ol>
        <li>Return all the rows (axis=0) from index 1000 to 1009 using regex with filter function only HINT: data.filter(axis=0,regex="")</li>
    <li>Return all the Even indexes from the above range using regex with filter function only i.e. OUTPUT:[1000, 1002, 1004, 1006, 1008]</li>
        <li>Reading Task: data.filter(regex="year"),data.filter(regex="(?i)year") Explain the difference between the two statements in a print statement and explain the effect its having HINT: Use data.filter(regex="artist") to get some insight first.</li>
    <li>Reading Task: Research and Write difference between Dataframe and Data Series in Pandas according to your understanding.</li></ol>

</ol>
Remember axis=0 represents the row operations and axis=1 represents the column wise operations. When axis=0 the filter is applied on the row indexes only i.e. the id column. Hint for home task need to use [ ) , | , ^ , ( , $ ] operators in correct order. For understanding and testing your regex go to : https://regexr.com/

### Data Aggregation(Task 1)

In [1]:
import pandas as pd

In [3]:
data = pd.read_csv('artwork_sample.csv')
data

id accession_number          artist artistRole  artistId  \
0  1035           A00001   Blake, Robert     artist        38   
1  1036           A00002   Blake, Robert     artist        38   
2  1037           A00003   Blake, Robert     artist        38   
3  1038           A00004   Blake, Robert     artist        38   
4  1039           A00005  Blake, William     artist        39   
5  1040           A00006  Blake, William     artist        39   
6  1041           A00007  Blake, William     artist        39   
7  1042           A00008  Blake, William     artist        39   
8  1043           A00009  Blake, William     artist        39   
9  1044           A00010  Blake, William     artist        39   

                                               title                dateText  \
0  A Figure Bowing before a Seated Old Man with h...          date not known   
1  Two Drawings of Frightened Figures, Probably f...          date not known   
2  The Preaching of Warning. Verso: An Old Man En...                 ?c.1785   
3     Six Drawings of Figures with Outstretched Arms          date not known   
4  The Circle of the Lustful: Francesca da Rimini...  1826–7, reprinted 1892   
5      Ciampolo the Barrator Tormented by the Devils  1826–7, reprinted 1892   
6                        The Baffled Devils Fighting  1826–7, reprinted 1892   
7  The Six-Footed Serpent Attacking Agnolo Brunel...  1826–7, reprinted 1892   
8                 The Serpent Attacking Buoso Donati  1826–7, reprinted 1892   
9                 The Pit of Disease: The Falsifiers  1826–7, reprinted 1892   

                                              medium  \
0  Watercolour, ink, chalk and graphite on paper....   
1                                  Graphite on paper   
2        Graphite on paper. Verso: graphite on paper   
3                                  Graphite on paper   
4                            Line engraving on paper   
5                            Line engraving on paper   
6                            Line engraving on paper   
7                            Line engraving on paper   
8                            Line engraving on paper   
9                            Line engraving on paper   

                                          creditLine    year  acquisitionYear  \
0                Presented by Mrs John Richmond 1922     NaN             1922   
1                Presented by Mrs John Richmond 1922     NaN             1922   
2                Presented by Mrs John Richmond 1922  1785.0             1922   
3                Presented by Mrs John Richmond 1922     NaN             1922   
4  Purchased with the assistance of a special gra...  1826.0             1919   
5  Purchased with the assistance of a special gra...  1826.0             1919   
6  Purchased with the assistance of a special gra...  1826.0             1919   
7  Purchased with the assistance of a special gra...  1826.0             1919   
8  Purchased with the assistance of a special gra...  1826.0             1919   
9  Purchased with the assistance of a special gra...  1826.0             1919   

              dimensions  width  height  depth units  inscription  \
0  support: 394 x 419 mm    394     419    NaN    mm          NaN   
1  support: 311 x 213 mm    311     213    NaN    mm          NaN   
2  support: 343 x 467 mm    343     467    NaN    mm          NaN   
3  support: 318 x 394 mm    318     394    NaN    mm          NaN   
4    image: 243 x 335 mm    243     335    NaN    mm          NaN   
5    image: 240 x 338 mm    240     338    NaN    mm          NaN   
6    image: 242 x 334 mm    242     334    NaN    mm          NaN   
7    image: 246 x 340 mm    246     340    NaN    mm          NaN   
8    image: 241 x 335 mm    241     335    NaN    mm          NaN   
9    image: 243 x 340 mm    243     340    NaN    mm          NaN   

   thumbnailCopyright                                       thumbnailUrl  \
0                 NaN  http://www.tate.org.uk/art/images/work/A

In [4]:
data.head()

id accession_number          artist artistRole  artistId  \
0  1035           A00001   Blake, Robert     artist        38   
1  1036           A00002   Blake, Robert     artist        38   
2  1037           A00003   Blake, Robert     artist        38   
3  1038           A00004   Blake, Robert     artist        38   
4  1039           A00005  Blake, William     artist        39   

                                               title                dateText  \
0  A Figure Bowing before a Seated Old Man with h...          date not known   
1  Two Drawings of Frightened Figures, Probably f...          date not known   
2  The Preaching of Warning. Verso: An Old Man En...                 ?c.1785   
3     Six Drawings of Figures with Outstretched Arms          date not known   
4  The Circle of the Lustful: Francesca da Rimini...  1826–7, reprinted 1892   

                                              medium  \
0  Watercolour, ink, chalk and graphite on paper....   
1                                  Graphite on paper   
2        Graphite on paper. Verso: graphite on paper   
3                                  Graphite on paper   
4                            Line engraving on paper   

                                          creditLine    year  acquisitionYear  \
0                Presented by Mrs John Richmond 1922     NaN             1922   
1                Presented by Mrs John Richmond 1922     NaN             1922   
2                Presented by Mrs John Richmond 1922  1785.0             1922   
3                Presented by Mrs John Richmond 1922     NaN             1922   
4  Purchased with the assistance of a special gra...  1826.0             1919   

              dimensions  width  height  depth units  inscription  \
0  support: 394 x 419 mm    394     419    NaN    mm          NaN   
1  support: 311 x 213 mm    311     213    NaN    mm          NaN   
2  support: 343 x 467 mm    343     467    NaN    mm          NaN   
3  support: 318 x 394 mm    318     394    NaN    mm          NaN   
4    image: 243 x 335 mm    243     335    NaN    mm          NaN   

   thumbnailCopyright                                       thumbnailUrl  \
0                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
1                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
2                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
3                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
4                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   

                                                 url  
0  http://www.tate.org.uk/art/artworks/blake-a-fi...  
1  http://www.tate.org.uk/art/artworks/blake-two-...  
2  http://www.tate.org.uk/art/artworks/blake-the-...  
3  http://www.tate.org.uk/art/artworks/blake-six-...  
4  http://www.tate.org.uk/art/artworks/blake-the-...

In [9]:
min_year = data.year.min()
print(min_year)

max_year = data.year.max()
print(max_year)

mean_year = data.year.mean()
print(mean_year)

std_year = data.year.std()
print(std_year)

1785.0
1826.0
1820.142857142857
15.496543393378316


In [10]:
list_year = [min_year, max_year, mean_year, std_year]
print(list_year)

[1785.0, 1826.0, 1820.142857142857, 15.496543393378316]


In [12]:
data.artist.sum()

'Blake, RobertBlake, RobertBlake, RobertBlake, RobertBlake, WilliamBlake, WilliamBlake, WilliamBlake, WilliamBlake, WilliamBlake, William'

In [11]:
min_artist = data.artist.min()
print(min_artist)

max_artist = data.artist.max()
print(max_artist)

Blake, Robert
Blake, William


In [13]:
# Mean and Std doesn't work on artist column 

mean_artist = data.artist.mean()
print(mean_artist)

std_artist = data.artist.std()
print(std_artist)



TypeError: Could not convert Blake, RobertBlake, RobertBlake, RobertBlake, RobertBlake, WilliamBlake, WilliamBlake, WilliamBlake, WilliamBlake, WilliamBlake, William to numeric

In [14]:
data.sum()

id                                                                10395
accession_number      A00001A00002A00003A00004A00005A00006A00007A000...
artist                Blake, RobertBlake, RobertBlake, RobertBlake, ...
artistRole            artistartistartistartistartistartistartistarti...
artistId                                                            386
title                 A Figure Bowing before a Seated Old Man with h...
dateText              date not knowndate not known?c.1785date not kn...
medium                Watercolour, ink, chalk and graphite on paper....
creditLine            Presented by Mrs John Richmond 1922Presented b...
year                                                              12741
acquisitionYear                                                   19202
dimensions            support: 394 x 419 mmsupport: 311 x 213 mmsupp...
width                                                              2821
height                                                          

In [61]:
agg_min = data.agg(min)

print(agg_min)

#minmum values or string in each column is printed below

id                                                                    1035
accession_number                                                    A00001
artist                                                       Blake, Robert
artistRole                                                          artist
artistId                                                                38
title                    A Figure Bowing before a Seated Old Man with h...
dateText                                            1826–7, reprinted 1892
medium                                                   Graphite on paper
creditLine                             Presented by Mrs John Richmond 1922
year                                                                  1785
acquisitionYear                                                       1919
dimensions                                             image: 240 x 338 mm
width                                                                  240
height                   

In [19]:
agg = data.agg(['min', 'max', 'mean', 'std'])

agg

id accession_number          artist artistRole   artistId  \
min   1035.00000           A00001   Blake, Robert     artist  38.000000   
max   1044.00000           A00010  Blake, William     artist  39.000000   
mean  1039.50000              NaN             NaN        NaN  38.600000   
std      3.02765              NaN             NaN        NaN   0.516398   

                                                  title  \
min   A Figure Bowing before a Seated Old Man with h...   
max   Two Drawings of Frightened Figures, Probably f...   
mean                                                NaN   
std                                                 NaN   

                    dateText  \
min   1826–7, reprinted 1892   
max           date not known   
mean                     NaN   
std                      NaN   

                                                 medium  \
min                                   Graphite on paper   
max   Watercolour, ink, chalk and graphite on paper....   
mean                                                NaN   
std                                                 NaN   

                                             creditLine         year  \
min                 Presented by Mrs John Richmond 1922  1785.000000   
max   Purchased with the assistance of a special gra...  1826.000000   
mean                                                NaN  1820.142857   
std                                                 NaN    15.496543   

      acquisitionYear             dimensions       width      height  depth  \
min       1919.000000    image: 240 x 338 mm  240.000000  213.000000    NaN   
max       1922.000000  support: 394 x 419 mm  394.000000  467.000000    NaN   
mean      1920.200000                    NaN  282.100000  351.500000    NaN   
std          1.549193                    NaN   55.562677   66.818577    NaN   

     units  inscription  thumbnailCopyright  \
min     mm          NaN                 NaN   
max     mm          NaN                 NaN   
mean   NaN          NaN                 NaN   
std    NaN          NaN                 NaN   

                                           thumbnailUrl  \
min   http://www.tate.org.uk/art/images/work/A/A00/A...   
max   http://www.tate.org.uk/art/images/work/A/A00/A...   
mean                                                NaN   
std                                                 NaN   

                                                    url  
min   http://www.tate.org.uk/art/artworks/blake-a-fi...  
max   http://www.tate.org.uk/art/artworks/blake-two-...  
mean                                                NaN  
std                                                 NaN

### Data Normalization (Task 2)

In [24]:
import pandas as pd

In [25]:
data = pd.read_csv('artwork_sample.csv')
data

id accession_number          artist artistRole  artistId  \
0  1035           A00001   Blake, Robert     artist        38   
1  1036           A00002   Blake, Robert     artist        38   
2  1037           A00003   Blake, Robert     artist        38   
3  1038           A00004   Blake, Robert     artist        38   
4  1039           A00005  Blake, William     artist        39   
5  1040           A00006  Blake, William     artist        39   
6  1041           A00007  Blake, William     artist        39   
7  1042           A00008  Blake, William     artist        39   
8  1043           A00009  Blake, William     artist        39   
9  1044           A00010  Blake, William     artist        39   

                                               title                dateText  \
0  A Figure Bowing before a Seated Old Man with h...          date not known   
1  Two Drawings of Frightened Figures, Probably f...          date not known   
2  The Preaching of Warning. Verso: An Old Man En...                 ?c.1785   
3     Six Drawings of Figures with Outstretched Arms          date not known   
4  The Circle of the Lustful: Francesca da Rimini...  1826–7, reprinted 1892   
5      Ciampolo the Barrator Tormented by the Devils  1826–7, reprinted 1892   
6                        The Baffled Devils Fighting  1826–7, reprinted 1892   
7  The Six-Footed Serpent Attacking Agnolo Brunel...  1826–7, reprinted 1892   
8                 The Serpent Attacking Buoso Donati  1826–7, reprinted 1892   
9                 The Pit of Disease: The Falsifiers  1826–7, reprinted 1892   

                                              medium  \
0  Watercolour, ink, chalk and graphite on paper....   
1                                  Graphite on paper   
2        Graphite on paper. Verso: graphite on paper   
3                                  Graphite on paper   
4                            Line engraving on paper   
5                            Line engraving on paper   
6                            Line engraving on paper   
7                            Line engraving on paper   
8                            Line engraving on paper   
9                            Line engraving on paper   

                                          creditLine    year  acquisitionYear  \
0                Presented by Mrs John Richmond 1922     NaN             1922   
1                Presented by Mrs John Richmond 1922     NaN             1922   
2                Presented by Mrs John Richmond 1922  1785.0             1922   
3                Presented by Mrs John Richmond 1922     NaN             1922   
4  Purchased with the assistance of a special gra...  1826.0             1919   
5  Purchased with the assistance of a special gra...  1826.0             1919   
6  Purchased with the assistance of a special gra...  1826.0             1919   
7  Purchased with the assistance of a special gra...  1826.0             1919   
8  Purchased with the assistance of a special gra...  1826.0             1919   
9  Purchased with the assistance of a special gra...  1826.0             1919   

              dimensions  width  height  depth units  inscription  \
0  support: 394 x 419 mm    394     419    NaN    mm          NaN   
1  support: 311 x 213 mm    311     213    NaN    mm          NaN   
2  support: 343 x 467 mm    343     467    NaN    mm          NaN   
3  support: 318 x 394 mm    318     394    NaN    mm          NaN   
4    image: 243 x 335 mm    243     335    NaN    mm          NaN   
5    image: 240 x 338 mm    240     338    NaN    mm          NaN   
6    image: 242 x 334 mm    242     334    NaN    mm          NaN   
7    image: 246 x 340 mm    246     340    NaN    mm          NaN   
8    image: 241 x 335 mm    241     335    NaN    mm          NaN   
9    image: 243 x 340 mm    243     340    NaN    mm          NaN   

   thumbnailCopyright                                       thumbnailUrl  \
0                 NaN  http://www.tate.org.uk/art/images/work/A

In [26]:
data.head()

id accession_number          artist artistRole  artistId  \
0  1035           A00001   Blake, Robert     artist        38   
1  1036           A00002   Blake, Robert     artist        38   
2  1037           A00003   Blake, Robert     artist        38   
3  1038           A00004   Blake, Robert     artist        38   
4  1039           A00005  Blake, William     artist        39   

                                               title                dateText  \
0  A Figure Bowing before a Seated Old Man with h...          date not known   
1  Two Drawings of Frightened Figures, Probably f...          date not known   
2  The Preaching of Warning. Verso: An Old Man En...                 ?c.1785   
3     Six Drawings of Figures with Outstretched Arms          date not known   
4  The Circle of the Lustful: Francesca da Rimini...  1826–7, reprinted 1892   

                                              medium  \
0  Watercolour, ink, chalk and graphite on paper....   
1                                  Graphite on paper   
2        Graphite on paper. Verso: graphite on paper   
3                                  Graphite on paper   
4                            Line engraving on paper   

                                          creditLine    year  acquisitionYear  \
0                Presented by Mrs John Richmond 1922     NaN             1922   
1                Presented by Mrs John Richmond 1922     NaN             1922   
2                Presented by Mrs John Richmond 1922  1785.0             1922   
3                Presented by Mrs John Richmond 1922     NaN             1922   
4  Purchased with the assistance of a special gra...  1826.0             1919   

              dimensions  width  height  depth units  inscription  \
0  support: 394 x 419 mm    394     419    NaN    mm          NaN   
1  support: 311 x 213 mm    311     213    NaN    mm          NaN   
2  support: 343 x 467 mm    343     467    NaN    mm          NaN   
3  support: 318 x 394 mm    318     394    NaN    mm          NaN   
4    image: 243 x 335 mm    243     335    NaN    mm          NaN   

   thumbnailCopyright                                       thumbnailUrl  \
0                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
1                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
2                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
3                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
4                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   

                                                 url  
0  http://www.tate.org.uk/art/artworks/blake-a-fi...  
1  http://www.tate.org.uk/art/artworks/blake-two-...  
2  http://www.tate.org.uk/art/artworks/blake-the-...  
3  http://www.tate.org.uk/art/artworks/blake-six-...  
4  http://www.tate.org.uk/art/artworks/blake-the-...

In [31]:
x = data.columns[data.dtypes == "int64"]
print(x)

x = data.columns[data.dtypes == "float"]
print(x)

#printed those calculated the int and float values and then normalized

Index(['id', 'artistId', 'acquisitionYear', 'width', 'height'], dtype='object')
Index(['year', 'depth', 'inscription', 'thumbnailCopyright'], dtype='object')


In [49]:
a = (data['height']- data['height'].mean())/data['height'].std()
a

0    1.010198
1   -2.072777
2    1.728561
3    0.636051
4   -0.246937
5   -0.202040
6   -0.261903
7   -0.172108
8   -0.246937
9   -0.172108
Name: height, dtype: float64

In [37]:
minmax = (data['height']- data['height'].min())/(data['height'].max()-data['height'].min())
minmax

0    0.811024
1    0.000000
2    1.000000
3    0.712598
4    0.480315
5    0.492126
6    0.476378
7    0.500000
8    0.480315
9    0.500000
Name: height, dtype: float64

In [38]:
print(minmax.max())
print(minmax.min())

1.0
0.0


In [39]:
data['standardized_height'] = minmax 
data

id accession_number          artist artistRole  artistId  \
0  1035           A00001   Blake, Robert     artist        38   
1  1036           A00002   Blake, Robert     artist        38   
2  1037           A00003   Blake, Robert     artist        38   
3  1038           A00004   Blake, Robert     artist        38   
4  1039           A00005  Blake, William     artist        39   
5  1040           A00006  Blake, William     artist        39   
6  1041           A00007  Blake, William     artist        39   
7  1042           A00008  Blake, William     artist        39   
8  1043           A00009  Blake, William     artist        39   
9  1044           A00010  Blake, William     artist        39   

                                               title                dateText  \
0  A Figure Bowing before a Seated Old Man with h...          date not known   
1  Two Drawings of Frightened Figures, Probably f...          date not known   
2  The Preaching of Warning. Verso: An Old Man En...                 ?c.1785   
3     Six Drawings of Figures with Outstretched Arms          date not known   
4  The Circle of the Lustful: Francesca da Rimini...  1826–7, reprinted 1892   
5      Ciampolo the Barrator Tormented by the Devils  1826–7, reprinted 1892   
6                        The Baffled Devils Fighting  1826–7, reprinted 1892   
7  The Six-Footed Serpent Attacking Agnolo Brunel...  1826–7, reprinted 1892   
8                 The Serpent Attacking Buoso Donati  1826–7, reprinted 1892   
9                 The Pit of Disease: The Falsifiers  1826–7, reprinted 1892   

                                              medium  \
0  Watercolour, ink, chalk and graphite on paper....   
1                                  Graphite on paper   
2        Graphite on paper. Verso: graphite on paper   
3                                  Graphite on paper   
4                            Line engraving on paper   
5                            Line engraving on paper   
6                            Line engraving on paper   
7                            Line engraving on paper   
8                            Line engraving on paper   
9                            Line engraving on paper   

                                          creditLine    year  ...  \
0                Presented by Mrs John Richmond 1922     NaN  ...   
1                Presented by Mrs John Richmond 1922     NaN  ...   
2                Presented by Mrs John Richmond 1922  1785.0  ...   
3                Presented by Mrs John Richmond 1922     NaN  ...   
4  Purchased with the assistance of a special gra...  1826.0  ...   
5  Purchased with the assistance of a special gra...  1826.0  ...   
6  Purchased with the assistance of a special gra...  1826.0  ...   
7  Purchased with the assistance of a special gra...  1826.0  ...   
8  Purchased with the assistance of a special gra...  1826.0  ...   
9  Purchased with the assistance of a special gra...  1826.0  ...   

              dimensions width  height  depth  units inscription  \
0  support: 394 x 419 mm   394     419    NaN     mm         NaN   
1  support: 311 x 213 mm   311     213    NaN     mm         NaN   
2  support: 343 x 467 mm   343     467    NaN     mm         NaN   
3  support: 318 x 394 mm   318     394    NaN     mm         NaN   
4    image: 243 x 335 mm   243     335    NaN     mm         NaN   
5    image: 240 x 338 mm   240     338    NaN     mm         NaN   
6    image: 242 x 334 mm   242     334    NaN     mm         NaN   
7    image: 246 x 340 mm   246     340    NaN     mm         NaN   
8    image: 241 x 335 mm   241     335    NaN     mm         NaN   
9    image: 243 x 340 mm   243     340    NaN     mm         NaN   

   thumbnailCopyright                                       thumbnailUrl  \
0                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
1                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
2                 NaN  http://www.tate.org.uk/art/imag

In [40]:
data.head()

id accession_number          artist artistRole  artistId  \
0  1035           A00001   Blake, Robert     artist        38   
1  1036           A00002   Blake, Robert     artist        38   
2  1037           A00003   Blake, Robert     artist        38   
3  1038           A00004   Blake, Robert     artist        38   
4  1039           A00005  Blake, William     artist        39   

                                               title                dateText  \
0  A Figure Bowing before a Seated Old Man with h...          date not known   
1  Two Drawings of Frightened Figures, Probably f...          date not known   
2  The Preaching of Warning. Verso: An Old Man En...                 ?c.1785   
3     Six Drawings of Figures with Outstretched Arms          date not known   
4  The Circle of the Lustful: Francesca da Rimini...  1826–7, reprinted 1892   

                                              medium  \
0  Watercolour, ink, chalk and graphite on paper....   
1                                  Graphite on paper   
2        Graphite on paper. Verso: graphite on paper   
3                                  Graphite on paper   
4                            Line engraving on paper   

                                          creditLine    year  ...  \
0                Presented by Mrs John Richmond 1922     NaN  ...   
1                Presented by Mrs John Richmond 1922     NaN  ...   
2                Presented by Mrs John Richmond 1922  1785.0  ...   
3                Presented by Mrs John Richmond 1922     NaN  ...   
4  Purchased with the assistance of a special gra...  1826.0  ...   

              dimensions width  height  depth  units inscription  \
0  support: 394 x 419 mm   394     419    NaN     mm         NaN   
1  support: 311 x 213 mm   311     213    NaN     mm         NaN   
2  support: 343 x 467 mm   343     467    NaN     mm         NaN   
3  support: 318 x 394 mm   318     394    NaN     mm         NaN   
4    image: 243 x 335 mm   243     335    NaN     mm         NaN   

   thumbnailCopyright                                       thumbnailUrl  \
0                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
1                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
2                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
3                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
4                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   

                                                 url standardized_height  
0  http://www.tate.org.uk/art/artworks/blake-a-fi...            0.811024  
1  http://www.tate.org.uk/art/artworks/blake-two-...            0.000000  
2  http://www.tate.org.uk/art/artworks/blake-the-...            1.000000  
3  http://www.tate.org.uk/art/artworks/blake-six-...            0.712598  
4  http://www.tate.org.uk/art/artworks/blake-the-...            0.480315  

[5 rows x 21 columns]

### Data Transformations(Task 3)

In [41]:
import pandas as pd

In [42]:
data = pd.read_csv('artwork_sample.csv')
data

id accession_number          artist artistRole  artistId  \
0  1035           A00001   Blake, Robert     artist        38   
1  1036           A00002   Blake, Robert     artist        38   
2  1037           A00003   Blake, Robert     artist        38   
3  1038           A00004   Blake, Robert     artist        38   
4  1039           A00005  Blake, William     artist        39   
5  1040           A00006  Blake, William     artist        39   
6  1041           A00007  Blake, William     artist        39   
7  1042           A00008  Blake, William     artist        39   
8  1043           A00009  Blake, William     artist        39   
9  1044           A00010  Blake, William     artist        39   

                                               title                dateText  \
0  A Figure Bowing before a Seated Old Man with h...          date not known   
1  Two Drawings of Frightened Figures, Probably f...          date not known   
2  The Preaching of Warning. Verso: An Old Man En...                 ?c.1785   
3     Six Drawings of Figures with Outstretched Arms          date not known   
4  The Circle of the Lustful: Francesca da Rimini...  1826–7, reprinted 1892   
5      Ciampolo the Barrator Tormented by the Devils  1826–7, reprinted 1892   
6                        The Baffled Devils Fighting  1826–7, reprinted 1892   
7  The Six-Footed Serpent Attacking Agnolo Brunel...  1826–7, reprinted 1892   
8                 The Serpent Attacking Buoso Donati  1826–7, reprinted 1892   
9                 The Pit of Disease: The Falsifiers  1826–7, reprinted 1892   

                                              medium  \
0  Watercolour, ink, chalk and graphite on paper....   
1                                  Graphite on paper   
2        Graphite on paper. Verso: graphite on paper   
3                                  Graphite on paper   
4                            Line engraving on paper   
5                            Line engraving on paper   
6                            Line engraving on paper   
7                            Line engraving on paper   
8                            Line engraving on paper   
9                            Line engraving on paper   

                                          creditLine    year  acquisitionYear  \
0                Presented by Mrs John Richmond 1922     NaN             1922   
1                Presented by Mrs John Richmond 1922     NaN             1922   
2                Presented by Mrs John Richmond 1922  1785.0             1922   
3                Presented by Mrs John Richmond 1922     NaN             1922   
4  Purchased with the assistance of a special gra...  1826.0             1919   
5  Purchased with the assistance of a special gra...  1826.0             1919   
6  Purchased with the assistance of a special gra...  1826.0             1919   
7  Purchased with the assistance of a special gra...  1826.0             1919   
8  Purchased with the assistance of a special gra...  1826.0             1919   
9  Purchased with the assistance of a special gra...  1826.0             1919   

              dimensions  width  height  depth units  inscription  \
0  support: 394 x 419 mm    394     419    NaN    mm          NaN   
1  support: 311 x 213 mm    311     213    NaN    mm          NaN   
2  support: 343 x 467 mm    343     467    NaN    mm          NaN   
3  support: 318 x 394 mm    318     394    NaN    mm          NaN   
4    image: 243 x 335 mm    243     335    NaN    mm          NaN   
5    image: 240 x 338 mm    240     338    NaN    mm          NaN   
6    image: 242 x 334 mm    242     334    NaN    mm          NaN   
7    image: 246 x 340 mm    246     340    NaN    mm          NaN   
8    image: 241 x 335 mm    241     335    NaN    mm          NaN   
9    image: 243 x 340 mm    243     340    NaN    mm          NaN   

   thumbnailCopyright                                       thumbnailUrl  \
0                 NaN  http://www.tate.org.uk/art/images/work/A

In [43]:
data.height.transform(lambda x: x/10)

0    41.9
1    21.3
2    46.7
3    39.4
4    33.5
5    33.8
6    33.4
7    34.0
8    33.5
9    34.0
Name: height, dtype: float64

In [46]:
x = data.groupby('artist').transform('mean')['height']
x

0    373.25
1    373.25
2    373.25
3    373.25
4    337.00
5    337.00
6    337.00
7    337.00
8    337.00
9    337.00
Name: height, dtype: float64

In [53]:
print("Take the mean of the height of every artist and printed it according to the each artist and that why it's repeating ")

Take the mean of the height of every artist and printed it according to the each artist and that why it's repeating 


In [55]:
x = data[data.artist == 'Blake, William'].height.mean()
print(x, 'mm')

337.0 mm


In [62]:
data['mean_height_by_artist'] = data.groupby('artist').transform('mean')['height']
data

id accession_number          artist artistRole  artistId  \
0  1035           A00001   Blake, Robert     artist        38   
1  1036           A00002   Blake, Robert     artist        38   
2  1037           A00003   Blake, Robert     artist        38   
3  1038           A00004   Blake, Robert     artist        38   
4  1039           A00005  Blake, William     artist        39   
5  1040           A00006  Blake, William     artist        39   
6  1041           A00007  Blake, William     artist        39   
7  1042           A00008  Blake, William     artist        39   
8  1043           A00009  Blake, William     artist        39   
9  1044           A00010  Blake, William     artist        39   

                                               title                dateText  \
0  A Figure Bowing before a Seated Old Man with h...          date not known   
1  Two Drawings of Frightened Figures, Probably f...          date not known   
2  The Preaching of Warning. Verso: An Old Man En...                 ?c.1785   
3     Six Drawings of Figures with Outstretched Arms          date not known   
4  The Circle of the Lustful: Francesca da Rimini...  1826–7, reprinted 1892   
5      Ciampolo the Barrator Tormented by the Devils  1826–7, reprinted 1892   
6                        The Baffled Devils Fighting  1826–7, reprinted 1892   
7  The Six-Footed Serpent Attacking Agnolo Brunel...  1826–7, reprinted 1892   
8                 The Serpent Attacking Buoso Donati  1826–7, reprinted 1892   
9                 The Pit of Disease: The Falsifiers  1826–7, reprinted 1892   

                                              medium  \
0  Watercolour, ink, chalk and graphite on paper....   
1                                  Graphite on paper   
2        Graphite on paper. Verso: graphite on paper   
3                                  Graphite on paper   
4                            Line engraving on paper   
5                            Line engraving on paper   
6                            Line engraving on paper   
7                            Line engraving on paper   
8                            Line engraving on paper   
9                            Line engraving on paper   

                                          creditLine    year  ...  \
0                Presented by Mrs John Richmond 1922     NaN  ...   
1                Presented by Mrs John Richmond 1922     NaN  ...   
2                Presented by Mrs John Richmond 1922  1785.0  ...   
3                Presented by Mrs John Richmond 1922     NaN  ...   
4  Purchased with the assistance of a special gra...  1826.0  ...   
5  Purchased with the assistance of a special gra...  1826.0  ...   
6  Purchased with the assistance of a special gra...  1826.0  ...   
7  Purchased with the assistance of a special gra...  1826.0  ...   
8  Purchased with the assistance of a special gra...  1826.0  ...   
9  Purchased with the assistance of a special gra...  1826.0  ...   

              dimensions width  height  depth  units inscription  \
0  support: 394 x 419 mm   394     419    NaN     mm         NaN   
1  support: 311 x 213 mm   311     213    NaN     mm         NaN   
2  support: 343 x 467 mm   343     467    NaN     mm         NaN   
3  support: 318 x 394 mm   318     394    NaN     mm         NaN   
4    image: 243 x 335 mm   243     335    NaN     mm         NaN   
5    image: 240 x 338 mm   240     338    NaN     mm         NaN   
6    image: 242 x 334 mm   242     334    NaN     mm         NaN   
7    image: 246 x 340 mm   246     340    NaN     mm         NaN   
8    image: 241 x 335 mm   241     335    NaN     mm         NaN   
9    image: 243 x 340 mm   243     340    NaN     mm         NaN   

   thumbnailCopyright                                       thumbnailUrl  \
0                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
1                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
2                 NaN  http://www.tate.org.uk/art/imag

In [64]:
data.head()

id accession_number          artist artistRole  artistId  \
0  1035           A00001   Blake, Robert     artist        38   
1  1036           A00002   Blake, Robert     artist        38   
2  1037           A00003   Blake, Robert     artist        38   
3  1038           A00004   Blake, Robert     artist        38   
4  1039           A00005  Blake, William     artist        39   

                                               title                dateText  \
0  A Figure Bowing before a Seated Old Man with h...          date not known   
1  Two Drawings of Frightened Figures, Probably f...          date not known   
2  The Preaching of Warning. Verso: An Old Man En...                 ?c.1785   
3     Six Drawings of Figures with Outstretched Arms          date not known   
4  The Circle of the Lustful: Francesca da Rimini...  1826–7, reprinted 1892   

                                              medium  \
0  Watercolour, ink, chalk and graphite on paper....   
1                                  Graphite on paper   
2        Graphite on paper. Verso: graphite on paper   
3                                  Graphite on paper   
4                            Line engraving on paper   

                                          creditLine    year  ...  \
0                Presented by Mrs John Richmond 1922     NaN  ...   
1                Presented by Mrs John Richmond 1922     NaN  ...   
2                Presented by Mrs John Richmond 1922  1785.0  ...   
3                Presented by Mrs John Richmond 1922     NaN  ...   
4  Purchased with the assistance of a special gra...  1826.0  ...   

              dimensions width  height  depth  units inscription  \
0  support: 394 x 419 mm   394     419    NaN     mm         NaN   
1  support: 311 x 213 mm   311     213    NaN     mm         NaN   
2  support: 343 x 467 mm   343     467    NaN     mm         NaN   
3  support: 318 x 394 mm   318     394    NaN     mm         NaN   
4    image: 243 x 335 mm   243     335    NaN     mm         NaN   

   thumbnailCopyright                                       thumbnailUrl  \
0                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
1                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
2                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
3                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
4                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   

                                                 url mean_height_by_artist  
0  http://www.tate.org.uk/art/artworks/blake-a-fi...                373.25  
1  http://www.tate.org.uk/art/artworks/blake-two-...                373.25  
2  http://www.tate.org.uk/art/artworks/blake-the-...                373.25  
3  http://www.tate.org.uk/art/artworks/blake-six-...                373.25  
4  http://www.tate.org.uk/art/artworks/blake-the-...                337.00  

[5 rows x 21 columns]

### Data Filtering (Task 4)

In [2]:
import pandas as pd

In [3]:
data = pd.read_csv("artwork_data.csv", low_memory = False)
data

id accession_number              artist artistRole  artistId  \
0        1035           A00001       Blake, Robert     artist        38   
1        1036           A00002       Blake, Robert     artist        38   
2        1037           A00003       Blake, Robert     artist        38   
3        1038           A00004       Blake, Robert     artist        38   
4        1039           A00005      Blake, William     artist        39   
...       ...              ...                 ...        ...       ...   
69196  122960           T13865  P-Orridge, Genesis     artist     16646   
69197  122961           T13866  P-Orridge, Genesis     artist     16646   
69198  121181           T13867        Hatoum, Mona     artist      2365   
69199  112306           T13868       Creed, Martin     artist      2760   
69200  127035           T13869   Brunias, Agostino     artist      3824   

                                                   title  \
0      A Figure Bowing before a Seated Old Man with h...   
1      Two Drawings of Frightened Figures, Probably f...   
2      The Preaching of Warning. Verso: An Old Man En...   
3         Six Drawings of Figures with Outstretched Arms   
4      The Circle of the Lustful: Francesca da Rimini...   
...                                                  ...   
69196                        Larvae (from Tampax Romana)   
69197                   Living Womb (from Tampax Romana)   
69198                                      Present Tense   
69199          Work No. 227: The lights going on and off   
69200                   Dancing Scene in the West Indies   

                     dateText  \
0              date not known   
1              date not known   
2                     ?c.1785   
3              date not known   
4      1826–7, reprinted 1892   
...                       ...   
69196                    1975   
69197                    1976   
69198                    1996   
69199                    2000   
69200                 1764–96   

                                                  medium  \
0      Watercolour, ink, chalk and graphite on paper....   
1                                      Graphite on paper   
2            Graphite on paper. Verso: graphite on paper   
3                                      Graphite on paper   
4                                Line engraving on paper   
...                                                  ...   
69196   Perspex, Wood, hairpiece, tampon and human blood   
69197  Wood, Perspex, plastic, photograph on paper, t...   
69198                               Soap and glass beads   
69199                                   Gallery lighting   
69200                                Oil paint on canvas   

                                              creditLine  year  \
0                    Presented by Mrs John Richmond 1922   NaN   
1                    Presented by Mrs John Richmond 1922   NaN   
2                    Presented by Mrs John Richmond 1922  1785   
3                    Presented by Mrs John Richmond 1922   NaN   
4      Purchased with the assistance of a special gra...  1826   
...                                                  ...   ...   
69196                 Transferred from Tate Archive 2012  1975   
69197                 Transferred from Tate Archive 2012  1976   
69198                     Presented by Tate Members 2013  1996   
69199  Purchased with funds provided by Tate Members,...  2000   
69200  Purchased with assistance from Tate Patrons an...  1764   

       acquisitionYear                           dimensions width height  \
0               1922.0                support: 394 x 419 mm   394    419   
1               1922.0                support: 311 x 213 mm   311    213   
2               1922.0                support: 343 x 467 mm   343    467   
3               1922.0                support: 318 x 394 mm   318    394   
4               1919.0                  image: 243 x 335 mm   243    335   
...                ...

In [4]:
data.head()

id accession_number          artist artistRole  artistId  \
0  1035           A00001   Blake, Robert     artist        38   
1  1036           A00002   Blake, Robert     artist        38   
2  1037           A00003   Blake, Robert     artist        38   
3  1038           A00004   Blake, Robert     artist        38   
4  1039           A00005  Blake, William     artist        39   

                                               title                dateText  \
0  A Figure Bowing before a Seated Old Man with h...          date not known   
1  Two Drawings of Frightened Figures, Probably f...          date not known   
2  The Preaching of Warning. Verso: An Old Man En...                 ?c.1785   
3     Six Drawings of Figures with Outstretched Arms          date not known   
4  The Circle of the Lustful: Francesca da Rimini...  1826–7, reprinted 1892   

                                              medium  \
0  Watercolour, ink, chalk and graphite on paper....   
1                                  Graphite on paper   
2        Graphite on paper. Verso: graphite on paper   
3                                  Graphite on paper   
4                            Line engraving on paper   

                                          creditLine  year  acquisitionYear  \
0                Presented by Mrs John Richmond 1922   NaN           1922.0   
1                Presented by Mrs John Richmond 1922   NaN           1922.0   
2                Presented by Mrs John Richmond 1922  1785           1922.0   
3                Presented by Mrs John Richmond 1922   NaN           1922.0   
4  Purchased with the assistance of a special gra...  1826           1919.0   

              dimensions width height  depth units inscription  \
0  support: 394 x 419 mm   394    419    NaN    mm         NaN   
1  support: 311 x 213 mm   311    213    NaN    mm         NaN   
2  support: 343 x 467 mm   343    467    NaN    mm         NaN   
3  support: 318 x 394 mm   318    394    NaN    mm         NaN   
4    image: 243 x 335 mm   243    335    NaN    mm         NaN   

  thumbnailCopyright                                       thumbnailUrl  \
0                NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
1                NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
2                NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
3                NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
4                NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   

                                                 url  
0  http://www.tate.org.uk/art/artworks/blake-a-fi...  
1  http://www.tate.org.uk/art/artworks/blake-two-...  
2  http://www.tate.org.uk/art/artworks/blake-the-...  
3  http://www.tate.org.uk/art/artworks/blake-six-...  
4  http://www.tate.org.uk/art/artworks/blake-the-...

In [1]:
data.filter(['id', 'artist']).head()

NameError: name 'data' is not defined

In [15]:
x = data.filter(axis = 0, regex = '100')[1:11]
x

id accession_number                          artist artistRole  \
1000  14704           A01004  Turner, Joseph Mallord William     artist   
1001  14705           A01005  Turner, Joseph Mallord William     artist   
1002  14706           A01006  Turner, Joseph Mallord William     artist   
1003  14707           A01007  Turner, Joseph Mallord William     artist   
1004  14708           A01008  Turner, Joseph Mallord William     artist   
1005  14709           A01009  Turner, Joseph Mallord William     artist   
1006  14710           A01010  Turner, Joseph Mallord William     artist   
1007  14711           A01011  Turner, Joseph Mallord William     artist   
1008  14712           A01012  Turner, Joseph Mallord William     artist   
1009  14713           A01013  Turner, Joseph Mallord William     artist   

      artistId                                              title dateText  \
1000       558                               Hedging and Ditching     1812   
1001       558     Hedging and Ditching, engraved by J.C. Easling     1812   
1002       558                                          River Wye     1812   
1003       558                    River Wye, engraved by W. Annis     1812   
1004       558            Chain of Alps from Grenoble to Chamberi     1812   
1005       558  Chain of Alps from Grenoble to Chamberi, engra...     1812   
1006       558                                       Mer de Glace     1812   
1007       558                                       Mer de Glace     1812   
1008       558                            Rivaux Abbey, Yorkshire     1812   
1009       558               Rivaux Abbey, engraved by Henry Dawe     1812   

                              medium  \
1000  Etching and engraving on paper   
1001  Etching and mezzotint on paper   
1002                  Print on paper   
1003  Etching and mezzotint on paper   
1004  Etching and engraving on paper   
1005  Etching and mezzotint on paper   
1006  Etching and engraving on paper   
1007  Etching and mezzotint on paper   
1008  Etching and engraving on paper   
1009  Etching and mezzotint on paper   

                                             creditLine  year  \
1000  Presented by A. Acland Allen through the Art F...  1812   
1001  Presented by A. Acland Allen through the Art F...  1812   
1002  Presented by A. Acland Allen through the Art F...  1812   
1003  Presented by A. Acland Allen through the Art F...  1812   
1004  Presented by A. Acland Allen through the Art F...  1812   
1005  Presented by A. Acland Allen through the Art F...  1812   
1006  Presented by A. Acland Allen through the Art F...  1812   
1007  Presented by A. Acland Allen through the Art F...  1812   
1008  Presented by A. Acland Allen through the Art F...  1812   
1009  Presented by A. Acland Allen through the Art F...  1812   

      acquisitionYear           dimensions width height  depth units  \
1000           1925.0  image: 184 x 259 mm   184    259    NaN    mm   
1001           1925.0  image: 186 x 262 mm   186    262    NaN    mm   
1002           1925.0  image: 183 x 264 mm   183    264    NaN    mm   
1003           1925.0  image: 184 x 265 mm   184    265    NaN    mm   
1004           1925.0  image: 178 x 259 mm   178    259    NaN    mm   
1005           1925.0  image: 178 x 261 mm   178    261    NaN    mm   
1006           1925.0  image: 178 x 253 mm   178    253    NaN    mm   
1007           1925.0  image: 179 x 257 mm   179    257    NaN    mm   
1008           1925.0  image: 182 x 265 mm   182    265    NaN    mm   
1009           1925.0  image: 183 x 266 mm   183    266    NaN    mm   

     inscription thumbnailCopyright  \
1000         NaN                NaN   
1001         NaN                NaN   
1002         NaN                NaN   
1003         NaN                NaN   
1004         NaN                NaN   
1005         NaN                NaN   
1006         NaN                NaN   
1007         NaN                NaN   
1008         NaN       

In [16]:
y = x.filter(regex = '1000|2|4|6|8', axis = 0)
y

id accession_number                          artist artistRole  \
1000  14704           A01004  Turner, Joseph Mallord William     artist   
1002  14706           A01006  Turner, Joseph Mallord William     artist   
1004  14708           A01008  Turner, Joseph Mallord William     artist   
1006  14710           A01010  Turner, Joseph Mallord William     artist   
1008  14712           A01012  Turner, Joseph Mallord William     artist   

      artistId                                    title dateText  \
1000       558                     Hedging and Ditching     1812   
1002       558                                River Wye     1812   
1004       558  Chain of Alps from Grenoble to Chamberi     1812   
1006       558                             Mer de Glace     1812   
1008       558                  Rivaux Abbey, Yorkshire     1812   

                              medium  \
1000  Etching and engraving on paper   
1002                  Print on paper   
1004  Etching and engraving on paper   
1006  Etching and engraving on paper   
1008  Etching and engraving on paper   

                                             creditLine  year  \
1000  Presented by A. Acland Allen through the Art F...  1812   
1002  Presented by A. Acland Allen through the Art F...  1812   
1004  Presented by A. Acland Allen through the Art F...  1812   
1006  Presented by A. Acland Allen through the Art F...  1812   
1008  Presented by A. Acland Allen through the Art F...  1812   

      acquisitionYear           dimensions width height  depth units  \
1000           1925.0  image: 184 x 259 mm   184    259    NaN    mm   
1002           1925.0  image: 183 x 264 mm   183    264    NaN    mm   
1004           1925.0  image: 178 x 259 mm   178    259    NaN    mm   
1006           1925.0  image: 178 x 253 mm   178    253    NaN    mm   
1008           1925.0  image: 182 x 265 mm   182    265    NaN    mm   

     inscription thumbnailCopyright  \
1000         NaN                NaN   
1002         NaN                NaN   
1004         NaN                NaN   
1006         NaN                NaN   
1008         NaN                NaN   

                                           thumbnailUrl  \
1000  http://www.tate.org.uk/art/images/work/A/A01/A...   
1002  http://www.tate.org.uk/art/images/work/A/A01/A...   
1004  http://www.tate.org.uk/art/images/work/A/A01/A...   
1006  http://www.tate.org.uk/art/images/work/A/A01/A...   
1008  http://www.tate.org.uk/art/images/work/A/A01/A...   

                                                    url  
1000  http://www.tate.org.uk/art/artworks/turner-hed...  
1002  http://www.tate.org.uk/art/artworks/turner-riv...  
1004  http://www.tate.org.uk/art/artworks/turner-cha...  
1006  http://www.tate.org.uk/art/artworks/turner-mer...  
1008  http://www.tate.org.uk/art/artworks/turner-riv...

In [17]:
data.filter(regex = 'year')

year
0       NaN
1       NaN
2      1785
3       NaN
4      1826
...     ...
69196  1975
69197  1976
69198  1996
69199  2000
69200  1764

[69201 rows x 1 columns]

In [18]:
data.filter(regex = '(?i)year')

year  acquisitionYear
0       NaN           1922.0
1       NaN           1922.0
2      1785           1922.0
3       NaN           1922.0
4      1826           1919.0
...     ...              ...
69196  1975           2013.0
69197  1976           2013.0
69198  1996           2013.0
69199  2000           2013.0
69200  1764           2013.0

[69201 rows x 2 columns]

In [19]:
print("(?i) is basically used to show the years  writeen in the word")

(?i) is basically used to show the years  writeen in the word


In [20]:
data.filter(regex = 'artist')

artist artistRole  artistId
0           Blake, Robert     artist        38
1           Blake, Robert     artist        38
2           Blake, Robert     artist        38
3           Blake, Robert     artist        38
4          Blake, William     artist        39
...                   ...        ...       ...
69196  P-Orridge, Genesis     artist     16646
69197  P-Orridge, Genesis     artist     16646
69198        Hatoum, Mona     artist      2365
69199       Creed, Martin     artist      2760
69200   Brunias, Agostino     artist      3824

[69201 rows x 3 columns]

In [22]:
print("Data Series is also called as array or a list because it has a mulptile rows and columns")

Data Series is also called as array or a list because it has a mulptile rows and columns
